kernel/kernel-api/src/main/java/io/delta/kernel/internal/skipping/StatsSchemaHelper.java

```diff
@@ -0,0 +1,275 @@
+     * the column exists, is a leaf column, and is of a skipping-eligible data-type.
+     */
+    public boolean isSkippingEligibleMinMaxColumn(Column column) {
+        return logicalToDataType.containsKey(column) &&
+            isSkippingEligibleDataType(logicalToDataType.get(column)) &&
+            // TODO for now we block using min/max columns of timestamps. JSON serialization
+            //   truncates to milliseconds. To safely use timestamp min/max stats we need to add
+            //   a millisecond to max statistics which requires time addition expression
+            !(logicalToDataType.get(column) instanceof TimestampType);
+    }
+
+    /**
+     * Returns true if the given column is skipping-eligible using null count statistics. This means
+     * the column exists and is a leaf column as we only collect stats for leaf columns.
+     */
+    public boolean isSkippingEligibleNullCountColumn(Column column) {
+        return logicalToPhysicalColumn.containsKey(column);
+    }
+
+    //////////////////////////////////////////////////////////////////////////////////
+    // Private helpers
+    //////////////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Given a logical column and a stats type returns the corresponding column in the statistics
+     * schema
+     */
+    private Column getStatsColumn(Column column, String statType) {
+        checkArgument(logicalToPhysicalColumn.containsKey(column),
+            String.format("%s is not a valid leaf column for data schema", column, dataSchema));
+        return getChildColumn(logicalToPhysicalColumn.get(column), statType);
+    }
+
+    /**
+     * Given a data schema returns a map of {logical column -> (physical column, data type)} for
+     * all leaf columns in the schema.
+     */
+    private Map<Column, Tuple2<Column, DataType>> getLogicalToPhysicalColumnAndDataType(
+            StructType dataSchema) {
+        Map<Column, Tuple2<Column, DataType>> result = new HashMap<>();
+        for (StructField field : dataSchema.fields()) {
+            if (field.getDataType() instanceof StructType) {
+                Map<Column, Tuple2<Column, DataType>> nestedCols =
+                    getLogicalToPhysicalColumnAndDataType((StructType) field.getDataType());
+                for (Column childLogicalCol : nestedCols.keySet()) {
+                    Column childPhysicalCol = nestedCols.get(childLogicalCol)._1;
+                    DataType childColDataType = nestedCols.get(childLogicalCol)._2;
+                    result.put(
+                        getChildColumn(childLogicalCol, field.getName()),
+                        new Tuple2<>(
+                            getChildColumn(childPhysicalCol, getPhysicalName(field)),
+                            childColDataType
+                        ));
+                }
+            } else {
+                result.put(
```

Comment:
qn: do we collect stats for map keys or values?

Reply:
We collect null_count stats (but not min/max). So we can generate skipping filters for expressions like `isNull` but not for others.